### PR TITLE
Support show enum values in Javadoc

### PR DESCRIPTION
**Pull request by haolingdong-msft**
Fix https://github.com/Azure/autorest.java/issues/1485

### CODE REVIEW

**Comment by weidongxu-microsoft** [comment]
Please think further. Did you cover all details?

File: javagen/src/main/java/com/azure/autorest/mapper/ClientParameterMapper.java [hunk 1]
```diff
@@ -79,6 +82,12 @@ public ClientMethodParameter map(Parameter parameter) {
         if (description == null || description.isEmpty()) {
             description = String.format("The %s parameter", name);
         }
+        // add allowed enum values
+        if (settings.isDataPlaneClient() && parameter.getProtocol().getHttp().getIn() != RequestParameterLocation.BODY && Mappers.getSchemaMapper().map(parameter.getSchema()) instanceof EnumType) {
+            description += ". ";
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/mapper/ClientParameterMapper.java:87]: What if description already ends with "."?

Can it be moved to buildAllowedEnumValues?

Why you need to map it `Mappers.getSchemaMapper().map(parameter.getSchema())`? Can you move all these to buildAllowedEnumValues?

File: javagen/src/main/java/com/azure/autorest/util/MethodUtil.java [hunk 2]
```diff
@@ -98,6 +101,25 @@ public static Request tryMergeBinaryRequests(List<Request> requests, Operation o
         return selectedRequest;
     }
 
+    /**
+     * build the string of allowed values for an enum type parameter
+     * @param parameter an enum type parameter
+     * @return the string of allowed values for this enum type parameter
+     */
+    public static String buildAllowedEnumValues(Parameter parameter) {
+        if (parameter.getSchema() == null || !(Mappers.getSchemaMapper().map(parameter.getSchema()) instanceof EnumType)) {
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/MethodUtil.java:110]: same here, why map?

At most, all these only require one map call.

File: javagen/src/main/java/com/azure/autorest/util/MethodUtil.java [hunk 3]
```diff
@@ -98,6 +101,25 @@ public static Request tryMergeBinaryRequests(List<Request> requests, Operation o
...
+            return "";
+        }
+        String res = "Allowed values: ";
+        EnumType enumType = (EnumType) Mappers.getSchemaMapper().map(parameter.getSchema());
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/MethodUtil.java:114]: Another map.

File: javagen/src/main/java/com/azure/autorest/util/MethodUtil.java [hunk 4]
```diff
@@ -98,6 +101,25 @@ public static Request tryMergeBinaryRequests(List<Request> requests, Operation o
...
+        List<ClientEnumValue> choices = enumType.getValues();
+        if (choices != null && !choices.isEmpty()) {
+            res += choices.stream().map(choice -> choice.getValue()).collect(Collectors.joining(", "));
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/MethodUtil.java:117]: Should we quote the value, if it is string?

What if enum contains ","?

What if enum contains "<"?

What if enum is "@abc"?

Did you handle all of them (maybe more) correctly?

**Review by weidongxu-microsoft — COMMENTED** [review]
I've added an additional CI check on javadoc of protocol-tests
https://github.com/Azure/autorest.java/pull/1507

File: javagen/src/main/java/com/azure/autorest/util/MethodUtil.java [hunk 5]
```diff
@@ -98,6 +102,37 @@ public static Request tryMergeBinaryRequests(List<Request> requests, Operation o
...
+        if (parameter.getSchema() == null || !(Mappers.getSchemaMapper().map(parameter.getSchema()) instanceof EnumType)) {
+            return description;
+        }
+        String res = "";
+        if (description.endsWith(".")) {
+            res = description + " Allowed values: ";
+        } else {
+            res = description + ". Allowed values: ";
+        }
+        EnumType enumType = (EnumType) Mappers.getSchemaMapper().map(parameter.getSchema());
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/MethodUtil.java:121]: there is still 2 map!

File: javagen/src/main/java/com/azure/autorest/util/MethodUtil.java [hunk 6]
```diff
@@ -98,6 +102,37 @@ public static Request tryMergeBinaryRequests(List<Request> requests, Operation o
...
+     * If the parameter is not enum type, return the description directly, otherwise append the string of allowed values to the description
+     * @param parameter a parameter
+     * @param description parameter description
+     * @return the description that appends the string of allowed values for enum type parameter
+     */
+    public static String appendAllowedEnumValuesForEnumType(Parameter parameter, String description) {
+        if (parameter.getSchema() == null || !(Mappers.getSchemaMapper().map(parameter.getSchema()) instanceof EnumType)) {
+            return description;
+        }
+        String res = "";
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/MethodUtil.java:115]: why init?

File: protocol-tests/swagger/enums.json [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+          {
+            "name": "query-non-required-enum",
+            "in": "query",
+            "required": false,
+            "type": "string",
+            "description": "non required enum with three values",
+            "enum": [
+              "test1",
+              "test2",
+              "test@</spec.,i`~!&*-al@char/>"
```
Comment (weidongxu-microsoft) [inline protocol-tests/swagger/enums.json:62]: in some corner case, "@abc" could fail javadoc, but let's ignore it for now.

Comment (haolingdong-msft) [reply protocol-tests/swagger/enums.json:62]: Javadoc looks like this:
image

Comment (weidongxu-microsoft) [reply protocol-tests/swagger/enums.json:62]: And put this to a required param as well.

**Review by weidongxu-microsoft — COMMENTED** [review]
Make it non-draft when you are ready.

Maybe I am the only one that ever ready draft PR...

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by haolingdong-msft** [comment]
@srnagar Could you please take a look at this pr?
